packages/cli-kit/src/plugins/extension.ts

```diff
@@ -0,0 +1,43 @@
+import {FanoutHookFunction} from '../plugins.js'
+import {ExtendableError} from '../error.js'
+import {JsonMap} from '../json.js'
+
+export type ExtensionErrorType = 'unknown'
+export class ExtensionError extends ExtendableError {
+  type: ExtensionErrorType
+  constructor(type: ExtensionErrorType, message?: string) {
+    super(message)
+    this.type = type
+  }
+}
+
+/**
+ * Extension Plugins types
+ *
+ * Any plugin that provides extension definitions should implement `defineExtensionSpec` and `defineExtensionPoint`
+ */
+export interface HookReturnPerExtensionPlugin {
+  extension_spec: {
+    options: {[key: string]: never}
+    pluginReturns: {
+      [pluginName: string]: JsonMap
+    }
+  }
+  extension_point: {
+    options: {[key: string]: never}
+    pluginReturns: {
+      [pluginName: string]: JsonMap
+    }
+  }
+}
+
+export type ExtensionSpecFunction = FanoutHookFunction<'extension_spec', ''>
+export type ExtensionPointFunction = FanoutHookFunction<'extension_point', ''>
+
+export const defineExtensionSpec = (input: JsonMap): ExtensionSpecFunction => {
+  return async () => input
+}
+
+export const defineExtensionPoint = (input: JsonMap): ExtensionPointFunction => {
+  return async () => input
+}
```

Comment:
We might need to move this to the `app` package, so that it can use the extension types:

```ts
// app/src/plugins/extensions.ts

export const defineExtensionSpec = (input: LocalExtensionSpec): ExtensionSpecFunction => {
  return async () => input
}
```

and a plugin would import `app` instead of `cli-kit`, because the plugin is an extension of `app`

```ts
import { LocalExtensionSpec, defineExtensionSpec } from "@shopify/app/plugins/extensions";
```

Let's discuss this tomorrow :)